hopper/flash_fwd_launch_template.h

```diff
@@ -106,148 +152,465 @@ void run_flash_fwd(Flash_fwd_params &params, cudaStream_t stream) {
+      } else if (params.num_splits <= 16) {
+        kernel_combine = (void *) flash::combine_attn_seqk_parallel<FinalOutputType, ElementAccum, kHeadDim, kBlockM, 4, Is_even_K, Flash_fwd_params>;
+        smem_size_combine = sizeof(flash::SharedStorageLSE<float, Shape<Int<16>, Int<kBlockM+1>>, Shape<Int<16>>>);
+      } else if (params.num_splits <= 32) {
+        kernel_combine = (void *) flash::combine_attn_seqk_parallel<FinalOutputType, ElementAccum, kHeadDim, kBlockM, 5, Is_even_K, Flash_fwd_params>;
+        smem_size_combine = sizeof(flash::SharedStorageLSE<float, Shape<Int<32>, Int<kBlockM+1>>, Shape<Int<32>>>);
+      } else if (params.num_splits <= 64) {
+        kernel_combine = (void *) flash::combine_attn_seqk_parallel<FinalOutputType, ElementAccum, kHeadDim, kBlockM, 6, Is_even_K, Flash_fwd_params>;
+        smem_size_combine = sizeof(flash::SharedStorageLSE<float, Shape<Int<64>, Int<kBlockM+1>>, Shape<Int<64>>>);
+      } else if (params.num_splits <= 128) {
+        kernel_combine = (void *) flash::combine_attn_seqk_parallel<FinalOutputType, ElementAccum, kHeadDim, kBlockM, 7, Is_even_K, Flash_fwd_params>;
+        smem_size_combine = sizeof(flash::SharedStorageLSE<float, Shape<Int<128>, Int<kBlockM+1>>, Shape<Int<128>>>);
+      } else {
+        // don't support > 128 splits
+        return;
+      }
+      if (smem_size_combine >= 48 * 1024) {
+        CHECK_CUDA(cudaFuncSetAttribute(kernel_combine, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_size_combine));
+      }
+      dim3 grid_combine((params.b * params.h * params.seqlen_q + kBlockM - 1) / kBlockM);
+      dim3 block_dims_combine(128);
+      dim3 cluster_dims_combine(1, 1, 1);
+      cutlass::ClusterLaunchParams launch_params_combine{
+          grid_combine, block_dims_combine, cluster_dims_combine, smem_size_combine, stream};
+      cutlass::launch_kernel_on_cluster(launch_params_combine, kernel_combine, params);
+      CHECK_CUDA_KERNEL_LAUNCH();
+    }
 }
 
 template<typename T>
 void run_mha_fwd_hdim64(Flash_fwd_params &params, cudaStream_t stream) {
     constexpr static int Headdim = 64;
+    // constexpr static bool UseCluster = false;
+
     BOOL_SWITCH(params.is_causal, Is_causal, [&] {
-        BOOL_SWITCH(params.is_local, Is_local, [&] {
-            SEQLEN_SWITCH(params.cu_seqlens_q, Seqlen_traits, [&] {
-                run_flash_fwd<
-                    Flash_fwd_kernel_traits<Headdim, 192, 128, 16, 2, false, 1, T>, 
-                    Is_causal, Is_local && !Is_causal, Seqlen_traits
-                >(params, stream);
+      BOOL_SWITCH(params.is_local, Is_local, [&] {
+        MMA_3WG_SWITCH(params.seqlen_q, kNumMmaWGs, [&] {
+          SEQLEN_SWITCH(params.cu_seqlens_q, Seqlen_traits, [&] {
+            BOOL_SWITCH(params.num_splits > 1, Is_split, [&] {
+              run_flash_fwd<
+                  Flash_fwd_kernel_traits<Headdim, kNumMmaWGs * 64, 128, 4 + kNumMmaWGs * 4,
```

Comment:
Adjusting Q tile size would make it more efficient for small seq len, thanks for the fix~